src/Nethermind/Nethermind.Consensus/Processing/CensorshipDetector/CensorshipDetector.cs

```diff
@@ -0,0 +1,228 @@
+    private readonly ILogger _logger;
+    private readonly ICensorshipDetectorConfig _censorshipDetectorConfig;
+    private readonly Dictionary<AddressAsKey, Transaction?> _poolAddressCensorshipBestTx = [];
+    private readonly LruCache<BlockNumberHash, BlockCensorshipInfo> _potentiallyCensoredBlocks;
+    private readonly WrapAroundArray<BlockNumberHash> _censoredBlocks;
+    private const int _cacheSize = 64;
+
+    public CensorshipDetector(
+        ITxPool txPool,
+        IComparer<Transaction> comparer,
+        IBlockProcessor blockProcessor,
+        ILogManager logManager,
+        ICensorshipDetectorConfig censorshipDetectorConfig)
+    {
+        _txPool = txPool;
+        _comparer = comparer;
+        _blockProcessor = blockProcessor;
+        _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+        _censorshipDetectorConfig = censorshipDetectorConfig;
+
+        if (_censorshipDetectorConfig.AddressesForCensorshipDetection is not null)
+        {
+            foreach (string hexString in _censorshipDetectorConfig.AddressesForCensorshipDetection)
+            {
+                if (Address.TryParse(hexString, out Address address))
+                {
+                    _poolAddressCensorshipBestTx[address!] = null;
+                }
+            }
+        }
+
+        _potentiallyCensoredBlocks = new(_cacheSize, _cacheSize, "potentiallyCensoredBlocks");
+        _censoredBlocks = new(_cacheSize);
+        _blockProcessor.BlockProcessing += OnBlockProcessing;
+    }
+
+    private void OnBlockProcessing(object? sender, BlockEventArgs e)
+    {
+        UInt256 baseFee = e.Block.BaseFeePerGas;
+
+        IEnumerable<Transaction> poolBestTransactions = _txPool.GetBestTxOfEachSender();
+        foreach (Transaction tx in poolBestTransactions)
+        {
+            if (tx.To is not null && tx.GasBottleneck > baseFee && _poolAddressCensorshipBestTx.TryGetValue(tx.To!, out Transaction? bestTx))
+            {
+                if (bestTx is null)
+                {
+                    _poolAddressCensorshipBestTx[tx.To!] = tx;
+                }
+                else if (_comparer.Compare(bestTx, tx) > 0)
+                {
+                    _poolAddressCensorshipBestTx[tx.To!] = tx;
+                }
+            }
+        }
+
+        Task.Run(() => Cache(e.Block));
+    }
+
+    private void Cache(Block block)
+    {
+        Transaction bestTxInBlock = block.Transactions[0];
+        Transaction worstTxInBlock = block.Transactions[0];
+        HashSet<AddressAsKey> blockAddressCensorshipDetectorHelper = [];
+        // Number of unique addresses specified by the user for censorship detection, to which txs are sent in the block
+        long blockCensorshipDetectionUniqueAddressCount = 0;
+
+        foreach (Transaction tx in block.Transactions)
+        {
+            if (!tx.SupportsBlobs)
+            {
+                if (_poolAddressCensorshipBestTx.ContainsKey(tx.To!)
```

Comment:
I think we could compare with transaction that is not includable?

Reply:
in `_poolAddressCensorshipBestTx` there are only includable txs, it was checked when building this collection, in line 66 (`tx.GasBottleneck > baseFee`)

Reply:
for non-ready txs gas bottleneck is 0, so yes, tx readiness is guaranteed by `tx.GasBottleneck > baseFee`

Reply:
Right, but @Arindam2407 this should land in comments as it is a bit cryptic